client/v3/watch.go

```diff
@@ -880,12 +880,13 @@ func (w *watchGrpcStream) serveSubstream(ws *watcherStream, resumec chan struct{
 				}
 			} else {
 				// current progress of watch; <= store revision
-				nextRev = wr.Header.Revision
+				nextRev = wr.Header.Revision + 1
```

Comment:
I haven't read this code deeply... but please make sure the 'fragment'  field is properly interpreted:

https://github.com/etcd-io/etcd/blob/586eaccc4d28d4c2af507eaef0051d21a7e4b04c/api/etcdserverpb/rpc.proto#L835

I assume that in case of fragmented response, there might be multiple responses with the same revision.

Reply:
It's a good question. The fragments have already been merged before the merged response being dispatched to the `serveSubstream`. The rough workflow is as below,
1. Firstly the client side merges all fragments if needed. See [watch.go#L644-L648](https://github.com/etcd-io/etcd/blob/586eaccc4d28d4c2af507eaef0051d21a7e4b04c/client/v3/watch.go#L644-L648);
2. dispatch the response. See [watch.go#L651](https://github.com/etcd-io/etcd/blob/586eaccc4d28d4c2af507eaef0051d21a7e4b04c/client/v3/watch.go#L651);
3. The `serveSubstream` forwards the response to the subscriber. See [watch.go#L845](https://github.com/etcd-io/etcd/blob/586eaccc4d28d4c2af507eaef0051d21a7e4b04c/client/v3/watch.go#L845);

Reply:
Er... this is going to break Kubernetes, which manually does this already.

Reply:
agreed, please don't change the meaning of the data returned to offset by one

Reply:
Actually this may be correct -- look at line 887, which is the normal case. It adds one.

Reply:
I agree with @lavalamp . 

@logicalhan @liggitt Could you share the thought why K8s manually do this? I guess it's just because etcd did not fix it?

Reply:
is `wr.Header.Revision == wr.Events[len(wr.Events)-1].Kv.ModRevision` in the normal case?

Reply:
here
https://github.com/etcd-io/etcd/blob/586eaccc4d28d4c2af507eaef0051d21a7e4b04c/client/v3/watch.go#L886-L888

Reply:
@logicalhan was this the spot you were thinking of where Kubernetes adds one to the revision?

Saving `resp.Header.Revision`:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go#L209

Taking a revision to watch from: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go#L105-L109

Starting a watch one revision higher than the requested or last returned revision:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go#L236

Reply:
Note this PR is fixing the etcd clientv3. My understanding is that it doesn't impact K8s at all. When etcdserver crashes, the etcd clientv3 will automatically re-send the watch request with the `nextRev`; this process is transparent to K8s. 

From client perspective, K8s remember the last rev, and manually adding 1 is totally client's behavior, and obviously it's also correct.

Reply:
This is the type of change that seems a bit crazy to me without unit tests that demonstrate what is currently broken and what this fixes.

Reply:
> @logicalhan was this the spot you were thinking of where Kubernetes adds one to the revision?

Yeah.

Reply:
> This is the type of change that seems a bit crazy to me without unit tests that demonstrate what is currently broken and what this fixes.

We have test cases in separate PRs.

Reply:
ok, if this doesn't change the revisions returned to clients, and aligns the zero-event progress-notification handling calculation of `nextRev`, I think that makes sense

Reply:
> ok, if this doesn't change the revisions returned to clients

Right. It doesn't anything being returned to client (`again, etcd clientv3's user.`)

Reply:
I would expect test changes to come in the same PR as the fix -- like did this not break any unit tests? That would be very surprising. (but I'm not familiar with etcd practice)

Reply:
> I would expect test changes to come in the same PR as the fix -- like did this not break any unit tests? That would be very surprising. (but I'm not familiar with etcd practice)

Right. I am planning to add a test case today.

Reply:
I was thinking to depend on the [generic linearizablity test case](https://github.com/etcd-io/etcd/pull/15257), but it seems it makes more sense to add a specific test case to reproduce this issue.

Reply:
That there isn't a unit test exercising this line is probably most of the problem...